dbms/src/Storages/KVStore/FFI/JointThreadAllocInfo.cpp

```diff
@@ -0,0 +1,184 @@
+        {
+            std::unique_lock l(memory_allocation_mut);
+            m.insert_or_assign(tname, ThreadInfoJealloc());
+        }
+        break;
+    }
+    case ReportThreadAllocateInfoType::Remove:
+    {
+        std::unique_lock l(memory_allocation_mut);
+        m.erase(tname);
+        break;
+    }
+    case ReportThreadAllocateInfoType::AllocPtr:
+    {
+        std::shared_lock l(memory_allocation_mut);
+        if (value == 0)
+            return;
+        auto it = m.find(tname);
+        if unlikely (it == m.end())
+        {
+            return;
+        }
+        it->second.allocated_ptr = value;
+        break;
+    }
+    case ReportThreadAllocateInfoType::DeallocPtr:
+    {
+        std::shared_lock l(memory_allocation_mut);
+        if (value == 0)
+            return;
+        auto it = m.find(tname);
+        if unlikely (it == m.end())
+        {
+            return;
+        }
+        it->second.deallocated_ptr = value;
+        break;
+    }
+    }
+}
+
+void JointThreadInfoJeallocMap::reportThreadAllocInfoForKVStore(
+    std::string_view thdname,
+    ReportThreadAllocateInfoType type,
+    uint64_t value)
+{
+    reportThreadAllocInfoImpl(kvstore_map, thdname, type, value);
+}
+
+static const std::unordered_set<std::string> RECORD_WHITE_LIST_THREAD_PREFIX = {"ReadIndexWkr"};
+
+void JointThreadInfoJeallocMap::recordThreadAllocInfoForKVStore()
+{
+    std::shared_lock l(memory_allocation_mut);
+    std::unordered_map<std::string, uint64_t> agg_allocate;
+    std::unordered_map<std::string, uint64_t> agg_deallocate;
+    for (const auto & [k, v] : kvstore_map)
+    {
+        auto agg_thread_name = getThreadNameAggPrefix(k);
+        // Some thread may have shorter lifetime, we can't use this timed task here to upgrade.
+        if (RECORD_WHITE_LIST_THREAD_PREFIX.contains(agg_thread_name))
+        {
+            agg_allocate[agg_thread_name] += v.allocated();
+            agg_deallocate[agg_thread_name] += v.deallocated();
+        }
+    }
+    for (const auto & [k, v] : agg_allocate)
+    {
+        auto & tiflash_metrics = TiFlashMetrics::instance();
+        tiflash_metrics.setProxyThreadMemory("alloc_" + k, v);
+    }
+    for (const auto & [k, v] : agg_deallocate)
+    {
+        auto & tiflash_metrics = TiFlashMetrics::instance();
+        tiflash_metrics.setProxyThreadMemory("dealloc_" + k, v);
+    }
+}
+
+void JointThreadInfoJeallocMap::reportThreadAllocBatchForKVStore(
+    std::string_view name,
+    ReportThreadAllocateInfoBatch data)
+{
+    // Many threads have empty name, better just not handle.
+    if (name.empty())
+        return;
+    // TODO(jemalloc-trace) Could be costy.
+    auto k = getThreadNameAggPrefix(name);
+    auto & tiflash_metrics = TiFlashMetrics::instance();
+    tiflash_metrics.setProxyThreadMemory("alloc_" + k, data.alloc);
+    tiflash_metrics.setProxyThreadMemory("dealloc_" + k, data.dealloc);
+}
+
+void JointThreadInfoJeallocMap::stopThreadAllocInfo()
```

Comment:
We should call it in `ContextShared::shutdown` explicitly to make sure the monitoring_thread is stopped before `TiFlashMetrics::instance` is released

Reply:
Then I shall move joint_memory_allocation_map from Context to ContextShared?

Reply:
Seems TiFlashMetrics::instance() is a singleton, I think maybe it actually overlives Context
```
TiFlashMetrics & TiFlashMetrics::instance()
{
    static TiFlashMetrics inst; // Instantiated on first use.
    return inst;
}
```

Reply:
> Then I shall move joint_memory_allocation_map from Context to ContextShared?

Yes, `Context` is a session level instance, the long-live instances should be in `ContextShared`